homeassistant/components/apple_tv/config_flow.py

```diff
@@ -0,0 +1,438 @@
+
+
+class AppleTVConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Apple TV."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_PUSH
+
+    @staticmethod
+    @callback
+    def async_get_options_flow(config_entry):
+        """Get options flow for this handler."""
+        return AppleTVOptionsFlow(config_entry)
+
+    def __init__(self):
+        """Initialize a new AppleTVConfigFlow."""
+        self.scan_result = None
+        self.atv = None
+        self.protocol = None
+        self.pairing = None
+        self.credentials = {}  # Protocol -> credentials
+
+    async def async_step_invalid_credentials(self, info):
+        """Handle initial step when updating invalid credentials."""
+        await self.async_set_unique_id(info.get(CONF_IDENTIFIER))
+
+        # pylint: disable=no-member # https://github.com/PyCQA/pylint/issues/3167
+        self.context["title_placeholders"] = {"name": info.get(CONF_NAME)}
+
+        # pylint: disable=no-member # https://github.com/PyCQA/pylint/issues/3167
+        self.context["identifier"] = self.unique_id
+        return await self.async_step_reconfigure()
+
+    async def async_step_reconfigure(self, user_input=None):
+        """Inform user that reconfiguration is about to start."""
+        if user_input is not None:
+            return await self.async_find_device_wrapper(
+                self.async_begin_pairing, allow_exist=True
+            )
+
+        return self.async_show_form(step_id="reconfigure")
+
+    async def async_step_user(self, user_input=None):
+        """Handle the initial step."""
+        # Be helpful to the user and look for devices
+        if self.scan_result is None:
+            self.scan_result, _ = await device_scan(None, self.hass.loop)
+
+        errors = {}
+        default_suggestion = self._prefill_identifier()
+        if user_input is not None:
+            await self.async_set_unique_id(user_input[CONF_IDENTIFIER])
+            try:
+                await self.async_find_device()
+                return await self.async_step_confirm()
+            except DeviceNotFound:
+                errors["base"] = "device_not_found"
+            except DeviceAlreadyConfigured:
+                errors["base"] = "device_already_configured"
+            except exceptions.NoServiceError:
+                errors["base"] = "no_usable_service"
+            except Exception:  # pylint: disable=broad-except
+                _LOGGER.exception("Unexpected exception")
+                errors["base"] = "unknown"
+
+            # Use whatever the user entered as default value
+            default_suggestion = self.unique_id
+
+        return self.async_show_form(
+            step_id="user",
+            data_schema=vol.Schema(
+                {vol.Required(CONF_IDENTIFIER, default=default_suggestion): str}
+            ),
+            errors=errors,
+            description_placeholders={"devices": self._devices_str()},
```

Comment:
I think you should not propose a `devices_str` on the description but instead propose a list on your schema, so it will display a select to the user. Mor user friendly.

Exemples:
- #28968 config_flow.py `_show_trusted_device_form`
- #32704 config_flow.py `_show_setup_form` `CONF_API_VERSION`

Reply:
I still want to allow for manual input (as some people have problems with discovery, but works fine if they enter an IP address), which would require me to add both a drop down menu and a text box. Or combine through multiple steps. I agree that this is not optimal, but due to discovery working so randomly I still believe this is the "best" middle way to not make implementation too complex. It can either way be improved in the future.

Reply:
In deconz integration I offer the path to the user to select manual config, but also as a fallback if nothing is discovered https://github.com/home-assistant/core/blob/d5193e64de6ce6a4460003d5f47351a92f35bba3/homeassistant/components/deconz/config_flow.py#L112